Final/ThreadPool.h

```diff
@@ -0,0 +1,100 @@
+#ifndef THREAD_POOL_H
+#define THREAD_POOL_H
+
+#include <vector>
+#include <queue>
+#include <memory>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <future>
+#include <functional>
+#include <stdexcept>
+
+//https://github.com/progschj/ThreadPool
+
+class ThreadPool {
```

Comment:
let's move this thread pool into our `dlpf` namespace under a new `async` subnamespace. let's also rename it `thread_pool` for consistency, and move the implementations into a .cpp for cleanliness. you'll find many c++ libraries are exported as "header only" (i.e. they put all the code in the .h file) because they're easy to link in and compile into an existing project, but in practice this doesn't really scale well because you can't really compile components separately.

thread_pool.h
```
#ifndef THREAD_POOL_H
#define THREAD_POOL_H

#include <vector>
#include <queue>
#include <memory>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <future>
#include <functional>
#include <stdexcept>

//https://github.com/progschj/ThreadPool

namespace dlpf::async {
	class thread_pool {
		public:
	    thread_pool(size_t);
	    template<class F, class... Args>
	    auto enqueue(F&& f, Args&&... args) 
	        -> std::future<typename std::result_of<F(Args...)>::type>;
	    ~thread_pool();
		private:
		    // need to keep track of threads so we can join them
		    std::vector< std::thread > workers;
		    // the task queue
		    std::queue< std::function<void()> > tasks;

		    // synchronization
		    std::mutex queue_mutex;
		    std::condition_variable condition;
		    bool stop;
	};
} // dlpf::async

#endif // THREAD_POOL_H
```

thread_pool.cpp
```
// the constructor just launches some amount of workers
dlpf::async::thread_pool::thread_pool(size_t threads)
    :   stop(false)
{
    for(size_t i = 0;i<threads;++i)
        workers.emplace_back(
            [this]
            {
                for(;;)
                {
                    std::function<void()> task;

                    {
                        std::unique_lock<std::mutex> lock(this->queue_mutex);
                        this->condition.wait(lock,
                            [this]{ return this->stop || !this->tasks.empty(); });
                        if(this->stop && this->tasks.empty())
                            return;
                        task = std::move(this->tasks.front());
                        this->tasks.pop();
                    }

                    task();
                }
            }
        );
}

// the destructor joins all threads
dlpf::async::thread_pool::~thread_pool()
{
    {
        std::unique_lock<std::mutex> lock(queue_mutex);
        stop = true;
    }
    condition.notify_all();
    for(std::thread &worker: workers)
        worker.join();
}

// add new work item to the pool
template<class F, class... Args>
auto dlpf::async::thread_pool::enqueue(F&& f, Args&&... args) 
    -> std::future<typename std::result_of<F(Args...)>::type>
{
    using return_type = typename std::result_of<F(Args...)>::type;

    auto task = std::make_shared< std::packaged_task<return_type()> >(
            std::bind(std::forward<F>(f), std::forward<Args>(args)...)
        );

    std::future<return_type> res = task->get_future();
    {
        std::unique_lock<std::mutex> lock(queue_mutex);

        // don't allow enqueueing after stopping the pool
        if(stop)
            throw std::runtime_error("enqueue on stopped thread_pool");

        tasks.emplace([task](){ (*task)(); });
    }
    condition.notify_one();
    return res;
}
```
note how we remove all the `inline` annotations in the .cpp. `inline` is really only necessary when writing implementations in the .h, or for very short helper methods in the .cpp.